app/src/main/java/com/devindi/wallpaper/model/map/MapImageGenerator.kt

```diff
@@ -1,13 +1,91 @@
 package com.devindi.wallpaper.model.map
 
 import android.graphics.Bitmap
+import android.graphics.Canvas
+import com.devindi.wallpaper.misc.RectD
+import com.google.firebase.perf.metrics.AddTrace
+import org.osmdroid.util.MapTileIndex
+import timber.log.Timber
 
 /**
  * Generates bitmap by params (source, center coordinates, zoom, size)
  */
-class MapImageGenerator() {
+class MapImageGenerator(
+    private val tileProvider: SyncMapTileProvider,
+    private val sourceFactory: TileSourceFactory,
+    private val tileUtils: TileUtils
+) {
 
-    fun generate(): Bitmap {
-        throw NotImplementedError()
+    fun generate(mapSourceId: String, zoom: Int, centerLat: Double, centerLong: Double, width: Int, height: Int): Bitmap {
+        Timber.d("Generate bitmap: $centerLat, $centerLong, $width, $height")
+        val tileSource = sourceFactory.getTileSource(mapSourceId)
+        val tilesRect = tileUtils.getTilesRect(centerLat, centerLong, width, height, tileSource.tileSizePixels, zoom)
+        return generate(mapSourceId, tilesRect, zoom)
     }
-}
\ No newline at end of file
+
+    @AddTrace(name = "bitmap generation")
+    fun generate(sourceId: String, tiles: RectD, zoom: Int): Bitmap {
+        Timber.d("Generating image for %s", tiles)
+        val tileSource = sourceFactory.getTileSource(sourceId)
+
+        val maxIndex = Math.pow(2.0, zoom.toDouble()).toInt() - 1
+
+        val startX = Math.floor(tiles.left).toInt()
+        val endX = Math.min(maxIndex, Math.ceil(tiles.right).toInt())
+
+        val xRange = TilesRange(startX, endX, zoom)
+
+        val startY = Math.floor(tiles.top).toInt()
+        val endY = Math.min(maxIndex, Math.ceil(tiles.bottom).toInt())
+
+        val yRange = TilesRange(startY, endY, zoom)
+
+        val tileSize = tileSource.tileSizePixels
+        val left = calculateLeftOffset(tiles.left, tileSize)
+        val top = calculateTopOffset(tiles.top, tileSize)
+        val right = calculateRightOffset(tiles.right, tileSize)
+        val bottom = calculateBottomOffset(tiles.bottom, tileSize)
+
+        val target = Bitmap.createBitmap(
+            xRange.size() * tileSize - left - right,
+            yRange.size() * tileSize - top - bottom,
+            Bitmap.Config.RGB_565)
+        Timber.d("target size: ${target.height} ${target.width}")
+        val canvas = Canvas(target)
+        xRange.forEachIndexed { xIndex, xTile ->
+            yRange.forEachIndexed { yIndex, yTile ->
+                val pxLeft = xIndex * tileSize.toFloat() - left
+                val pxTop = yIndex * tileSize.toFloat() - top
+                val tileBitmap = tileProvider.getTile(tileSource, MapTileIndex.getTileIndex(zoom, xTile, yTile))
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/devindi/mapwall/blob/63cb55c913b0d690492f74a58d6634a0c41df234/app/src/main/java/com/devindi/wallpaper/model/map/MapImageGenerator.kt#L59"></span>Exceeded max line length (100) (cannot be auto-corrected)</td>
    </tr>
  </tbody>
</table>